Refactor card deck definitions

ChristophNiehoff opened the pull request:
This pull request refactors the definitions of card decks.
The card decks are defined in `cardDefinitions.ts`. All information (like number of suits, suit names etc.) are taken from this definition.
In the future, this will make it much easier to add new card decks to the game. This is a preliminary step for implementing #197 .

ghost91- commented on src/utils/cardDefinitions.ts:
```diff
@@ -0,0 +1,318 @@
+  return suitHoldingCard?.abbreviation ?? '';
+}
+
+export function getCardDisplayName(
```
The naming is a bit inconsistent here: Some other other functions also return something for a `Card` but don't have `Card` in there name.

ChristophNiehoff replied:
I think `getAbbreviation` was the only thing, not having `Card` in its name.
Will change it to `getCardAbbreviation`. Then all `Card`-related functions should have `Card` in their names.

ghost91- commented on src/utils/cardDefinitions.ts:
```diff
@@ -0,0 +1,318 @@
+
+export function getCardScore(
+  card: Card,
+  currentSuit: Suit | undefined,
```
What's the point of handling the `undefined` case here? Wouldn't it make more sense to handle that outside the function, because it seems like that would be an error case to me.

ghost91- commented on src/utils/constants.ts:
```diff
@@ -1,68 +1,38 @@
 export const MAX_NUMBER_PLAYERS = 9;
+export const DEFAULT_TURN_DURATION = 300;
 
+//TODO: Move this into dedicated server config file
```
we should either address this here or create an issue for it.

ChristophNiehoff replied:
Moved this into a dedicated config file.

ChristophNiehoff commented on src/game/__tests__/eop.test.js:
```diff
@@ -21,35 +17,14 @@ describe('game', () => {
 
   const startingPlayer = players['0'].getState().ctx.currentPlayer;
 
-  it('should shuffle cards correctly', () => {
```
It is still needed, but the test -as is was- simply did not fit anymore the code for shuffling the cards. 
I think this belongs to `// TODO: Testing!!!`

ChristophNiehoff replied:
Meanwhile, tests have been added ;)

ChristophNiehoff commented on src/game/moves.ts:
```diff
@@ -18,7 +19,7 @@ export function toggleModal(
     hasPlayerPassed(G, ctx) ||
     (G.threat.modal && G.threat.owner !== ctx.playerID) ||
-    G.suit === ''
+    !G.suit
```
Agreed.
Added a ticket for added such an eslint rule.

ChristophNiehoff commented on src/game/moves.ts:
```diff
@@ -266,8 +267,7 @@ export function draw(
   // only update the suit if no suit exists
-  if (suit === '')
-    suit = card.substr(0, 1) as 'A' | 'B' | 'C' | 'D' | 'E' | 'T';
+  if (!suit) suit = card.substr(0, 1) as Suit;
```
agreed

ChristophNiehoff commented on src/game/utils.ts:
```diff
@@ -111,6 +60,49 @@ export function setupGame(ctx: Ctx, setupData?: SetupData): GameState {
   };
 }
 
+// TODO: Testing!!!
```
I'll this to my ongoing ticket about adding Elevation of Privacy.

ChristophNiehoff commented on src/game/utils.ts:
```diff
@@ -111,6 +60,49 @@ export function setupGame(ctx: Ctx, setupData?: SetupData): GameState {
+function getPlayerHoldingStartingCard(
+  handsPerPlayers: Card[][],
+  startingCard: Card,
+): keyof Card[][] {
```
I wasn't aware of that.
Will change it to number.

ChristophNiehoff commented:
@ghost91-  Added a further commit containing a bunch of tests

ghost91- commented:
@ChristophNiehoff The tests look good! what about the few remaining comments?

ChristophNiehoff commented:
@ghost91- Closed or addressed the remaining comments.
The latest commit introduced a lot of noise as apparently the ports are used in an awful lot of locations. I'll create a ticket to clean this up at some point.

ghost91- commented:
@ChristophNiehoff approved :)
